torch/utils/_sympy/value_ranges.py

```diff
@@ -261,31 +261,36 @@ def exp(x):
 
     @staticmethod
     def log(x):
+        if x.lower <= 0:
+            return ValueRanges.unknown()
         return ValueRanges.increasing_map(
             x, lambda y: -sympy.oo if y <= 0 else sympy.log(y)
         )
 
     @staticmethod
     def sqrt(x):
+        if x.lower < 0:
+            return ValueRanges.unknown()
         return ValueRanges.increasing_map(x, sympy.sqrt)
 
-    @staticmethod
-    def pow(a, b):
-        def is_integer(val):
-            return (
-                isinstance(val, int)
-                or (isinstance(val, float) and val == int(val))
-                or (hasattr(val, "is_integer") and val.is_integer)
-            )
-
+    @classmethod
+    def pow(cls, a, b):
         a = ValueRanges.wrap(a)
         b = ValueRanges.wrap(b)
-        if a.lower < 0 and not is_integer(b.lower):
-            # The function is not defined
-            return ValueRanges.unknown()
-        elif 0 in a and b.lower <= 0:
+        if a.lower == a.upper and b.lower == b.upper:
+            r = a.lower ** b.lower
+            if r == sympy.zoo:
+                return ValueRanges.unknown()
+            return ValueRanges.wrap(r)
+        elif b.lower == b.upper and b.lower >= 0:
+            i = ValueRanges.wrap(1)
+            for _ in range(b.lower):
+                i = cls.mul(i, a)
+            return i
+        else:
+            # This is fairly difficult to analyze, so give up for anything
+            # complicated
```

Comment:
Right, I see we were missing a couple cases here, yep.

Reply:
For future reference, it's probably enough to analyse the signs of a and b separately and have those 4 cases, but yeah, probably not worth the effort.

Reply:
PRs welcome! The problem I had was that you also have different regimes for a < 1 vs > 1 vs = 1. PAIN